core/py/src/braintrust_core/json.py

```diff
@@ -0,0 +1,69 @@
+import dataclasses
+import json
+from typing import Any, Dict, Optional
+
+
+class JSONSerializationError(Exception):
+    pass
+
+
+def check_json_serializable(obj: Any) -> None:
+    try:
+        _ = json.dumps(obj)
+    except Exception as e:
+        raise JSONSerializationError from e
+
+
+def is_json_serializable(obj: Any) -> bool:
+    try:
+        check_json_serializable(obj)
+        return True
+    except JSONSerializationError:
+        return False
+
+
+def dump_object(obj: Any) -> Any:
+    """Convert an object to a JSON-serializable object if it is an instance of `dataclasses.dataclass` or `pydantic.BaseModel`."""
+    if dataclasses.is_dataclass(obj) and not isinstance(obj, type):
+        return dataclasses.asdict(obj)
```

Comment:
I wonder if we want to recursively invoke `dump_object` on the result of this, in case the dataclass contains a pydantic model or something. Maybe that's too crazy.